pkg/volume/csi/nodeinfomanager/nodeinfomanager.go

```
@@ -0,0 +1,461 @@
+	// Concurrent calls to AddNodeInfo() is allowed, but they should not be intertwined with calls
+	// to other methods in this interface.
+	AddNodeInfo(driverName string, driverNodeID string, topology *csipb.Topology) error
+
+	// Remove in the cluster node information from the CSI driver with the given name.
+	// Concurrent calls to RemoveNodeInfo() is allowed, but they should not be intertwined with calls
+	// to other methods in this interface.
+	RemoveNodeInfo(driverName string) error
+}
+
+// NewNodeInfoManager initializes nodeInfoManager
+func NewNodeInfoManager(
+	nodeName types.NodeName,
+	kubeClient kubernetes.Interface,
+	csiKubeClient csiclientset.Interface) Interface {
+	return &nodeInfoManager{
+		nodeName:      nodeName,
+		k8s:           kubeClient,
+		csiKubeClient: csiKubeClient,
+	}
+}
+
+// AddNodeInfo updates the node ID annotation in the Node object and CSIDrivers field in the
+// CSINodeInfo object. If the CSINodeInfo object doesn't yet exist, it will be created.
+// If multiple calls to AddNodeInfo() are made in parallel, some calls might receive Node or
+// CSINodeInfo update conflicts, which causes the function to retry the corresponding update.
+func (nim *nodeInfoManager) AddNodeInfo(driverName string, driverNodeID string, topology *csipb.Topology) error {
+	if driverNodeID == "" {
+		return fmt.Errorf("error adding CSI driver node info: driverNodeID must not be empty")
+	}
+
+	if topology == nil {
+		return fmt.Errorf("error adding CSI driver node info: topology must not be nil")
+	}
+
+	err := nim.updateNode(
+		updateNodeIDInNode(driverName, driverNodeID),
+		updateTopologyLabels(topology),
+	)
+	if err != nil {
+		return fmt.Errorf("error updating Node object with CSI driver node info: %v", err)
+	}
+
+	err = nim.updateCSINodeInfo(driverName, driverNodeID, topology)
+	if err != nil {
+		return fmt.Errorf("error updating CSINodeInfo object with CSI driver node info: %v", err)
```

Maybe instead of failing, just log a warning for now saying you could not update the object and that topology may not work (caller has to enable CRD in order for it to work). And leave a TODO, once CSINodeInfo object goes beta you can fail instead of log?

---

ACK will update. Currently if an error is returned, the registration handler proceeds to unregister the driver, and we definitely don't wan this if CRD creation is disabled

---

It would have to move to GA right, since a user can still disable the CRD in beta?